### ISSUE TITLE
Lost ID bug with redo & eval

### ISSUE THREAD

**Issue by brprice** [hackworthltd/primer#1056]
### Code of conduct

- [X] I've read the project's code of conduct, and agree to follow it.

### Don't report security issues here

- [X] I've read the project's [security policy](https://github.com/hackworthltd/primer/security/policy), and I do not believe this bug is a security issue.

### Is there an existing issue?

- [X] I've searched GitHub Issues, and I believe this is a new issue

### What happened?

Related to #59, but this one is fails on "redo" rather than "undo", is reproducable, and has a somewhat clear cause.

I came across a situation where i had a program which seemed to be sensible and running actions seemed to work. `undo` also worked, but `redo` did not, or at least not fully.

### What should have happened?

Undo and redo to both work (when not at the end of the timeline) and to be inverses

### Which package(s) or aspect(s) of the project are the source of the issue?

primer

### Developers, please provide the output of the bug report script

```
The bug report script is running in a Nix shell.

System:
uname is:  Linux sweevo 6.1.34 #1-NixOS SMP PREEMPT_DYNAMIC Wed Jun 14 09:15:34 UTC 2023 x86_64 GNU/Linux

Nix:
nix-build path is:  /run/current-system/sw/bin/nix-build
nix-build version is:  nix-build (Nix) 2.17.0pre20230602_3c4d678
nix-command is enabled
flakes are enabled

direnv:
direnv is not in the PATH.

git:
git path is:  /run/current-system/sw/bin/git
git version is:  git version 2.40.1
git revision is:  e11e53df55f9370a13eb3678d49e377a0a7a80b3

make:
make path is:  /nix/store/nnis9ph3wj866bbs0lalbf22i5bq8yry-gnumake-4.4.1/bin/make
make version is:  GNU Make 4.4.1
Built for x86_64-pc-linux-gnu
Copyright (C) 1988-2023 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Haskell:
ghc path is:  /nix/store/6la7zx4xgj9s1n8m1kw6y176w2cdblg5-ghc-shell-for-packages-ghc-9.4.5-env/bin/ghc
ghc version is:  The Glorious Glasgow Haskell Compilation System, version 9.4.5
cabal path is:  /nix/store/qh2ws461gzm0l3lwzg6y1g83s4pvp8cd-cabal-install-exe-cabal-3.10.1.0/bin/cabal
cabal version is:  3.10.1.0

Sqitch:
sqitch path is:  /nix/store/7kcywgx1ijv4wvl8kv7y6m07fs68bsjn-sqitch-1.3.1/bin/sqitch
sqitch version is:  .sqitch-wrapped (App::Sqitch) v1.3.1
primer-sqitch path is:  /nix/store/4c5b0as5y22g8knjmpr54sa6jxqm4a3a-primer-sqitch-1.0/bin/primer-sqitch

SQLite:
sqlite3 path is:  /nix/store/f4wgc8q8v3alsykqlbh8qhmih55as2fs-sqlite-3.41.2-bin/bin/sqlite3
sqlite3 version is:  3.41.2 2023-03-22 11:56:21 0d1fc92f94cb6b76bffe3ec34d69cffde2924203304e8ffc4155597af0c191da

PostgreSQL:
psql path is:  /nix/store/czrrap8sqkkplqj6j8b8r97jca7zfcbm-postgresql-14.8/bin/psql
psql version is:  psql (PostgreSQL) 14.8

Colima:
colima path is:  /nix/store/0jlaik2cr2yx30z3w6hasijh8vm3hp3g-colima-0.4.6/bin/colima
colima version is:  colima version 0.4.6
git commit: 10377f3
```

### Do you have a reproducer?

- Start a new program (with prelude imported); I was using the primer-app frontend, but the problem is clearly in the backend.
- use the "apply function to arguments" in the term hole in `main`, and select `even`
- open a live eval window targeting `main` (I'm not sure whether it is important to leave it open, or if just triggering one eval is enough, and whether the eval should happen here or maybe some other point is sufficient to trigger the bug)
- insert the literal `4` into the hole, giving `main : ? ; main = even 4`
- press `undo`: it works, giving `main = even ?`
- press `redo`: it works, giving `main = even 4`
- (these last two can be repeated, if desired)
- press `undo` twice, giving `main = ?`
- press `redo`, it works, giving `main = even ?`
- press `redo`, it fails (nothing happens), and the error console complains about a lost ID
- note that `undo` and `redo` cycling between `main = ?` and `main = even ?` work, but any further `redo` fails.

### Any additional information?

I expect that this could affect undo as well, with some tweaks to the reproducer.

I suspect there is some problem where running the evaluator and doing program construction draw from the same pool of IDs, causing a desync when trying to replay

**Comment by brprice** [comment]
I have extended #973 to intersperse some evals in with the actions, and it quickly trips over some potential problems. However, I have not looked into these yet.

**Comment by dhess** [comment]
Yes, I think you might be on to something. I can reproduce similar behavior by building some fairly trivial functions, everything seems to be working, eval one or two of them; then create a new definition, start editing it, and now undo doesn't work.

**Comment by dhess** [comment]
This has been fixed, yes?

**Comment by brprice** [comment]
Yes, this has been fixed by #1086
